src/CodeViewer/Viewer.cpp

```diff
@@ -26,32 +27,57 @@ static const QColor kLineNumberBackgroundColor{50, 50, 50};
 static const QColor kLineNumberForegroundColor{189, 189, 189};
 static const QColor kTextEditBackgroundColor{30, 30, 30};
 static const QColor kTextEditForegroundColor{189, 189, 189};
+static const QColor kTitleBackgroundColor{30, 65, 89};
 static const QColor kHeatmapColor{Qt::red};
 
+static int StringWidthInPixels(const QFontMetrics& font_metrics, QString string) {
+  return font_metrics.horizontalAdvance(string);
+}
+
 int DetermineLineNumberWidthInPixels(const QFontMetrics& font_metrics, int max_line_number) {
-  return font_metrics.horizontalAdvance(QString::number(max_line_number));
+  return StringWidthInPixels(font_metrics, QString::number(max_line_number));
 }
 
-Viewer::Viewer(QWidget* parent) : QPlainTextEdit(parent), left_sidebar_widget_{this} {
-  UpdateLeftSidebarWidth();
-  QObject::connect(this, &QPlainTextEdit::blockCountChanged, this, &Viewer::UpdateLeftSidebarWidth);
+Viewer::Viewer(QWidget* parent)
+    : QPlainTextEdit(parent),
+      top_bar_widget_{this},
+      left_sidebar_widget_{this},
+      right_sidebar_widget_{this} {
+  UpdateBarsSize();
+  QObject::connect(this, &QPlainTextEdit::blockCountChanged, this, &Viewer::UpdateBarsSize);
+
+  QObject::connect(&top_bar_widget_, &PlaceHolderWidget::PaintEventTriggered, this,
+                   &Viewer::DrawTopWidget);
 
   QObject::connect(&left_sidebar_widget_, &PlaceHolderWidget::PaintEventTriggered, this,
                    &Viewer::DrawLineNumbers);
 
-  const auto update_viewport_area = [&](const QRect& rect, int dy) {
-    const auto update_caused_by_scroll = [](int dy) { return dy != 0; };
+  QObject::connect(&right_sidebar_widget_, &PlaceHolderWidget::PaintEventTriggered, this,
+                   &Viewer::DrawSampleCounters);
 
-    if (update_caused_by_scroll(dy)) {
+  const auto update_viewport_area = [&](const QRect& rect, int dy) {
+    bool update_caused_by_scroll = (dy != 0);
+    if (update_caused_by_scroll) {
       left_sidebar_widget_.scroll(0, dy);
+      right_sidebar_widget_.scroll(0, dy);
     } else {
+      QRect titles_bar = rect;
       QRect line_number_rect = rect;
+      QRect samples_info_rect = rect;
+
+      // We keep the (vertical) y-coordinates and adjust the horizontal x-coordinates for the each
+      // bar's area
+      titles_bar.setLeft(0);
+      titles_bar.setWidth(top_bar_widget_.sizeHint().width());
+      top_bar_widget_.update(line_number_rect);
```

Comment:
The title bar doesn't need to be updated at all, because a text change can't change anything here. I think you can remove it entirely.